### ISSUE TITLE
Corrupted Media  Files cause Video On Demand Crashes 

### ISSUE THREAD

**Issue by skawirat** [awslabs/video-on-demand-on-aws#44]
Hi @dscpinheiro,

We recently encountered an issue where the Video On Demand was crashing; and after investigation it turned out to be due an corrupted media file.

Would it be possible for you to suggest a quick fix so that rather than crashing it will send an error code and gracefully abort the job.

### Below is our high level understanding of the problem 

Normally when a file is handed over to VOD; it is passed onto mediainfo to extract information and JSON format output is redirected to a file.
Elsewhere in the code the JSON structure in file is iterated the media info is extracted. 

When a corrupted file goes through the above process mediainfo output indicates an incomplete file ( sets IsTruncated   to  Yes ). However the code that reads the MediaInfo does not handle the scenario where the JSON file does not contain required fields and crashes.

### Below are details of our investigation

error-handler cloud watch contains the following entries 

```
2019-11-18T20:25:41.195Z	c8c66424-aabf-464c-8e3a-746b673980c4	{ guid: 'fef1131d-0f55-4225-95b7-d95573ca08af',
  workflowStatus: 'Error',
  workflowErrorAt: 'V0d-R3v-mediainfo',
  errorMessage: 'TypeError: tracks.forEach is not a function',
  errorDetails: 'https://console.aws.amazon.com/cloudwatch/home?region=ap-southeast-2#logStream:group=/aws/lambda/V0d-R3v-mediainfo' }
```

Media Info Cloud watch contains the following entries

```
2019-11-18T20:25:40.334Z	9c940511-1ea9-4847-ac76-40a4103ae8c6	TypeError: tracks.forEach is not a function at MediaInfo.analyze (/var/task/lib/mediaInfo.js:425:14) at <anonymous> at process._tickDomainCallback (internal/process/next_tick.js:228:7)
2019-11-18T20:25:40.334Z	9c940511-1ea9-4847-ac76-40a4103ae8c6	TypeError: tracks.forEach is not a function
    at MediaInfo.analyze (/var/task/lib/mediaInfo.js:425:14)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
```

```
2019-11-18T20:25:42.627Z	9c940511-1ea9-4847-ac76-40a4103ae8c6	{"errorMessage":"tracks.forEach is not a function","errorType":"TypeError","stackTrace":["MediaInfo.analyze (/var/task/lib/mediaInfo.js:425:14)","<anonymous>","process._tickDomainCallback (internal/process/next_tick.js:228:7)"]}
2019-11-18T20:25:42.627Z	9c940511-1ea9-4847-ac76-40a4103ae8c6	
{
    "errorMessage": "tracks.forEach is not a function",
    "errorType": "TypeError",
    "stackTrace": [
        "MediaInfo.analyze (/var/task/lib/mediaInfo.js:425:14)",
        "<anonymous>",
        "process._tickDomainCallback (internal/process/next_tick.js:228:7)"
    ]
}
```

The actual crash seems to occur in mediainfo.js  at start of following for loop 

```
      tracks.forEach((track) => {
        switch (track.$.type) {
          case 'General':
            this.container = MediaInfo.parseGeneralAttributes(track);
            break;
          case 'Video':
            this.videoES.push(MediaInfo.parseVideoAttributes(track));
            break;
```

DynamoDB contained the following entry for the problem job
```
{
  "errorDetails": "https://console.aws.amazon.com/cloudwatch/home?region=ap-southeast-2#logStream:group=/aws/lambda/Optus-Sp0rt-V0d-R3v-mediainfo",
  "errorMessage": "TypeError: tracks.forEach is not a function",
  "guid": "fef1131d-0f55-4225-95b7-d95573ca08af",
  "workflowErrorAt": "V0d-R3v-mediainfo",
  "workflowStatus": "Error"
}
```

I am assuming you will be able to reproduce the issue with any invalid file.

Any suggestions for a quick fix  for this issue would be very much appreciated  

Thanks

Sam

**Comment by dscpinheiro** [comment]
Hi @skawirat,

I haven't tested the solution with a corrupted file, but I'll give it a try tomorrow.

The workflow only needs two properties from mediainfo (width and height), but I'm not really sure if it makes sense to continue if the video is invalid.

**Comment by skawirat** [comment]
Hi @dscpinheiro,

> >but I'm not really sure if it makes sense to continue if the video is invalid.

I agree there is no point trying to continue if the video is invalid.
If the  VOD system can send us a SNS with an error code/description prior to halting execution that would be great.

The current issue is since we receive no SNS notifications from VOD when it encounters a corrupt file; our front end applications have no knowledge of the status ( eg: is it progressing, is it error, is complete.etc) of submitted it submitted to VOD.

Thanks

Sam

**Comment by dscpinheiro** [comment]
Are you not getting a notification at all? That's odd, because the function that invokes mediainfo has a catch block: https://github.com/awslabs/video-on-demand-on-aws/blob/16bef0e010d0cd340546b65af0e0a0723c7b50e3/source/mediainfo/index.js#L39-L43

And the error handler does send a SNS notification: https://github.com/awslabs/video-on-demand-on-aws/blob/16bef0e010d0cd340546b65af0e0a0723c7b50e3/source/error-handler/index.js#L87-L92

I've tested in another part of the workflow (specifying a non-existent file), and I got the error email.

Could you confirm that the _error-handler_ function is being invoked? You should be able to see it in CloudWatch Logs.

**Comment by skawirat** [comment]
Thanks @dscpinheiro,

Until your clarification above we were under the impression we were not receiving SNS notifications for the scenario I explained.

However after double checking the SNS subscriptions we identified anissue from on our side where the SNS messages were not properly handled.

Since we now know that you are sending SNS notifications for this scenario we will do the needful from out side to handle that SNS message correctly.

Thank you very much for your prompt response..

Very much appreciated.

Sam
